core/new-gui/src/app/workspace/component/code-editor-dialog/code-editor-dialog.component.ts

```diff
@@ -1,57 +1,66 @@
+
+
+export const CODE_SAVE_ON_EDITION_DEBOUNCE_TIME_MS = 100;
 
 /**
  * CodeEditorDialogComponent is the content of the dialogue invoked by CodeareaCustomTemplateComponent.
  *
- * It contains an Ace editor which is inside a resizable mat-dialog-content. When the dialogue is invoked by
+ * It contains a Monaco editor which is inside a mat-dialog-content. When the dialogue is invoked by
  * the button in CodeareaCustomTemplateComponent, the data of the custom field (or empty String if no data)
- * will be sent to the Ace editor as its text. The dialogue can be closed with ESC key or by clicking on areas outside
- * the dialogue. Closing the dialogue will send the eidted contend back to the custom template field.
+ * will be sent to the Monaco editor as its text. The dialogue can be closed with ESC key or by clicking on areas outside
+ * the dialogue. Closing the dialogue will send the edited contend back to the custom template field.
  * @author Xiaozhen Liu
  */
 @Component({
   selector: 'texera-code-editor-dialog',
   templateUrl: './code-editor-dialog.component.html',
   styleUrls: ['./code-editor-dialog.component.scss']
 })
-export class CodeEditorDialogComponent implements OnInit {
-
-  aclOptions = {
-    enableBasicAutocompletion: true,
-    enableSnippets: true,
-    enableLiveAutocompletion: true,
-    maxLines: 40,
-    minLines: 20,
-    autoScrollEditorIntoView: false,
-    highlightActiveLine: true,
-    highlightSelectedWord: true,
-    highlightGutterLine: true,
-    animatedScroll: true
-  };
-
-  text: string;
+export class CodeEditorDialogComponent implements OnInit, OnDestroy {
+
+  editorOptions = { theme: 'vs-dark', language: 'python', fontSize: '11', automaticLayout: true};
+  code: string;
+  subscriptions: Subscription = new Subscription();
 
   constructor(
     private dialogRef: MatDialogRef<CodeEditorDialogComponent>,
-    @Inject(MAT_DIALOG_DATA) data: any) {
-      this.text = data;
-    }
+    @Inject(MAT_DIALOG_DATA) code: any,
+    private workflowActionService: WorkflowActionService
+  ) {
+    this.code = code;
+  }
 
   ngOnInit() {
-    this.dialogRef.keydownEvents().subscribe(event => {
-      if (event.key === 'Escape') {
-          this.onCancel();
-      }
-  });
-
-  this.dialogRef.backdropClick().subscribe(event => {
-      this.onCancel();
-  });
+    this.registerAutoSaveUponClosure();
+    this.registerAutoSaveUponEdition();
   }
 
-  onCancel(): void {
-    this.dialogRef.close(this.text);
-}
+  ngOnDestroy(): void {
+    this.subscriptions.unsubscribe();
+  }
+
+  private registerAutoSaveUponClosure() {
+    this.subscriptions.add(merge(
+      this.dialogRef.keydownEvents().filter(event => event.key === 'Escape'),
+      this.dialogRef.backdropClick()
+    ).subscribe(_ => this.dialogRef.close(this.code)));
+  }
 
+  private registerAutoSaveUponEdition() {
+    this.subscriptions.add(this.dialogRef.keydownEvents().debounceTime(CODE_SAVE_ON_EDITION_DEBOUNCE_TIME_MS).subscribe(_ => {
```

Comment:
I was wondering if keydown events are enough. For example, the user might right click to paste some text......

Reply:
good point. let me check

Reply:
changed to listen to `onModelChange`